### PR TITLE
Follow the target kernel's module hash algorithm

### LINKED ISSUE #266: sha512 signed modules not works
Modules signed with the dkms-3.0.8 are rejected and I end up at the prompt of initramfs. In my case zfs-modul. The modules are signed but not accepted. Neither with my own keys nor the ones created by dkms. I am using Debian SID

The reason for this can be found in the following line.

eval '"$sign_file" sha512 "$mok_signing_key" "$mok_certificate" "$built_module"'

Here I get hash_algo sha512 signed which doesn't work on my box and my
laptop. If I change this to sha256 secure-boot works as desired. Why
this is so I can not say, this is beyond my knowledge. Also if this is a
bug or not I can't say.

Sorry for my crap english :-)

...

### PR DESCRIPTION
**Pull request by xuzhen**
Fixes #266

### CODE REVIEW

File: run_test.sh [hunk 1]
```diff
@@ -380,6 +381,50 @@ EOF
...
+
+dkms_test.ko${mod_compression_ext}:
+Running module version sanity check.
+ - Original module
+   - No original module exists within this kernel
+ - Installation
+   - Installing to /lib/modules/${KERNEL_VER}/${expected_dest_loc}/
+depmod...
+EOF
+        run_with_expected_output sh -c "modinfo '/lib/modules/${KERNEL_VER}/${expected_dest_loc}/dkms_test.ko${mod_compression_ext}' | grep '^sig_hashalgo:' | sed 's/sig_hashalgo: *//'" << EOF
```
Comment (evelikov) [inline run_test.sh:422]: No need to filter out the "sig_hashalgo: " piece, just add it to the expectation below?

File: run_test.sh [hunk 2]
```diff
@@ -380,6 +381,50 @@ EOF
...
     fi
+
+    # If the key can't be extracted from module, the hash algorithm is also unknown
+    if [[ "${SIG_KEY// /}" != "sig_key:" ]] && grep -q '^CONFIG_MODULE_SIG_HASH=' "/lib/modules/${KERNEL_VER}/build/.config"; then
+        echo 'Building the test module using a different hash algorithm'
+        if grep -q '^CONFIG_MODULE_SIG_HASH="sha512"' "/lib/modules/${KERNEL_VER}/build/.config"; then
+            ALTER_HASH="sha256"
+        else
+            ALTER_HASH="sha512"
+        fi
```
Comment (evelikov) [inline run_test.sh:392]: Seems a bit odd to see "Building..." where the first thing immediately after is uninstalling.
Maybe move the above echo + if block, further down - just above the 'echo "CONFIG_..." > /tmp/...' line?

File: run_test.sh (hunk 1 above)
Comment (xuzhen) [reply run_test.sh]: OK

File: run_test.sh (hunk 2 above)
Comment (xuzhen) [reply run_test.sh]: The installing and uninstalling were removed. They were originally to workaround a bug. I'll open a new issue for that bug.

File: dkms.in [hunk 3]
```diff
@@ -953,6 +953,13 @@ prepare_signing()
         return
     fi
 
+    if [[ -f "${kernel_config}" ]]; then
+        sign_hash="$(grep "^CONFIG_MODULE_SIG_HASH=" "${kernel_config}" | sed 's/CONFIG_MODULE_SIG_HASH=//;s/"//g')"
+        sign_hash="${sign_hash:-sha512}"
```
Comment (evelikov) [inline dkms.in:958]: This piece is very odd - there is no single default sighash in the kernel. The default is picked based on what's built-in.

As such, I cannot see how the CONFIG option can be missing. Do you mind adding a note where you've seen this - "noticed with $distro $version, where $sighash is used but the CONFIG entry is missing".

That said, do you mind if we put your (good PRs) on hold for a day or two - would be nice to land https://github.com/dell/dkms/pull/262 without forcing the developer to rebase yet again.

Comment (evelikov) [reply dkms.in:958]: If the kernel lacks sig_hash option/support, perhaps we should skip the test all together?

Comment (xuzhen) [reply dkms.in:958]: > This piece is very odd - there is no single default sighash in the kernel. The default is picked based on what's built-in.
> As such, I cannot see how the CONFIG option can be missing. 

The kernel may be built without the module signing feature completely. `# CONFIG_MODULE_SIG_HASH is not set` in kconfig.

> Do you mind adding a note where you've seen this - "noticed with $distro $version, where $sighash is used but the CONFIG entry is missing".

OK

> That said, do you mind if we put your (good PRs) on hold for a day or two - would be nice to land #262 without forcing the developer to rebase yet again.

I don't mind. 

> If the kernel lacks sig_hash option/support, perhaps we should skip the test all together?

Currently dkms always signs the built modules, and there is no problem using signed modules on systems without module signing feature. The test is checking whether the hash algorithm of module signature is correct, not whether the system supports module signing. The test is necessary IMO.

Comment (evelikov) [reply dkms.in:959]: Thanks for bearing with my questions :bowing_man:

**Comment by mavoga** [comment]
I would have preferred

`sign_hash="$(grep "^CONFIG_MODULE_SIG_HASH=" "${kernel_config}" | cut -d= -f2 | tr -d '"')"`

or even

```
eval " $(grep "^CONFIG_MODULE_SIG_HASH=" "${kernel_config}")"
sign_hash="${CONFIG_MODULE_SIG_HASH:-sha512}"
unset CONFIG_MODULE_SIG_HASH
```

My 2 cents

**Comment by evelikov** [comment]
@mavoga I'm open for the former, but I would avoid `eval` unless strictly needed. PRs welcome o/

**Pull request merged**